File: src/IKVM.Runtime/Java/Externs/sun/security/provider/NativeSeedGenerator.cs
using System.Security.Cryptography;

namespace IKVM.Java.Externs.sun.security.provider
{

    /// <summary>
    /// Implements the native backend for 'NativeSeedGenerator'. This implementation uses the .NET RandomNumberGenerator.
    /// </summary>
    static class NativeSeedGenerator
    {

        static readonly RandomNumberGenerator rng = RandomNumberGenerator.Create();

        /// <summary>
        /// Implements the native method 'getSeedBytes'.
        /// </summary>
        /// <param name="result"></param>
        /// <returns></returns>
        public static bool nativeGenerateSeed(byte[] result)
        {
            rng.GetBytes(result);
            return true;
        }

    }

}

File: src/IKVM.Java/local/sun/security/provider/NativeSeedGenerator.java
package sun.security.provider;

import java.io.IOException;

/**
 * Seed generator for IKVM making use of native .NET features.
 *
 */
class NativeSeedGenerator extends SeedGenerator {

    /**
     * Create a new native seed generator instances.
     *
     * @exception IOException if native seeds are not available
     * on this platform.
     */
    NativeSeedGenerator(String seedFile) throws IOException {
        super();
    }

    /**
     * Native method to do the actual work.
     */
    private static native boolean nativeGenerateSeed(byte[] result);

    @Override
    void getSeedBytes(byte[] result) {
        if (nativeGenerateSeed(result) == false) {
            throw new InternalError("Unexpected failure generating seed");
        }
    }

}
